src/solvers/dgsem_p4est/subcell_limiters_2d.jl

```diff
@@ -0,0 +1,257 @@
+            var_secondary = u[variable, i_secondary, j_secondary, secondary_element]
+
+            var_min[i_primary, j_primary, primary_element] = min(var_min[i_primary,
+                                                                         j_primary,
+                                                                         primary_element],
+                                                                 var_secondary)
+            var_max[i_primary, j_primary, primary_element] = max(var_max[i_primary,
+                                                                         j_primary,
+                                                                         primary_element],
+                                                                 var_secondary)
+
+            var_min[i_secondary, j_secondary, secondary_element] = min(var_min[i_secondary,
+                                                                               j_secondary,
+                                                                               secondary_element],
+                                                                       var_primary)
+            var_max[i_secondary, j_secondary, secondary_element] = max(var_max[i_secondary,
+                                                                               j_secondary,
+                                                                               secondary_element],
+                                                                       var_primary)
+
+            # Increment primary element indices
+            i_primary += i_primary_step
+            j_primary += j_primary_step
+            i_secondary += i_secondary_step
+            j_secondary += j_secondary_step
+        end
+    end
+
+    # Calc bounds at physical boundaries
+    calc_bounds_twosided_boundary!(var_min, var_max, variable, u, t,
+                                   boundary_conditions,
+                                   mesh, equations, dg, cache)
+
+    return nothing
+end
+
+@inline function calc_bounds_twosided_boundary!(var_min, var_max, variable, u, t,
+                                                boundary_conditions::BoundaryConditionPeriodic,
+                                                mesh::P4estMesh{2},
+                                                equations, dg, cache)
+    return nothing
+end
+
+@inline function calc_bounds_twosided_boundary!(var_min, var_max, variable, u, t,
+                                                boundary_conditions,
+                                                mesh::P4estMesh{2},
+                                                equations, dg, cache)
+    (; boundary_condition_types, boundary_indices) = boundary_conditions
+    (; contravariant_vectors) = cache.elements
+
+    (; boundaries) = cache
+    index_range = eachnode(dg)
+
+    foreach(enumerate(boundary_condition_types)) do (i, boundary_condition)
```

Comment:
This is likely not type stable... not sure if it is a problem per se, but it might impact performance and cause additional allocations.

Reply:
I'll take a look. Thanks.

Reply:
Thanks. I fixed it in [3017b54](https://github.com/trixi-framework/Trixi.jl/pull/1954/commits/3017b541be874095a66fabcbea3d237407775ad3).
Don't know in which branch/PR this got lost, but I actually was aware of those allocations and implemented a version of `foreach_enumerate` even before it was used for the coupled simulations. Well, never mind :smile: